init custom gradle tooling model

glefloch opened the pull request:
This branch aims to expose a custom model containing project structure. This model is generating using the gradle API. 

This allows gradle users to:
* run/debug tests from IDE
* run/debug main `Quarkus.run` (simple / multi-module)

This branch closes: 
* #8737, #9738, #8950 related to run tests
* #10602, #10123 related to run main

@aloubyansky can you have a look? this should not evolve more (excepts with review comment and some test additions)

glefloch committed:
init custom gradle tooling model

aloubyansky commented on devtools/gradle/src/main/java/io/quarkus/gradle/AppModelGradleResolver.java:
```diff
@@ -95,32 +69,17 @@ public void relink(AppArtifact appArtifact, Path localPath) throws AppModelResol
-                        && appArtifact.getGroupId().equals(a.getModuleVersion().getId().getGroup())) {
-                    appArtifact.setPath(a.getFile().toPath());
+            for (io.quarkus.gradle.workspace.model.Dependency dependency : model.getAppDependencies()) {
+                // TODO there used to be a check on appArtifact.getType()
```
Is it not available in the model?

aloubyansky commented on devtools/gradle/src/main/java/io/quarkus/gradle/AppModelGradleResolver.java:
```diff
@@ -95,32 +69,17 @@ public void relink(AppArtifact appArtifact, Path localPath) throws AppModelResol
-                        && appArtifact.getType().equals(a.getType())
-                        && appArtifact.getGroupId().equals(a.getModuleVersion().getId().getGroup())) {
-                    appArtifact.setPath(a.getFile().toPath());
+            for (io.quarkus.gradle.workspace.model.Dependency dependency : model.getAppDependencies()) {
```
This is actually very limiting, since the scope is limited to the app model only. I'll need to check whether the devtools could be affected.

aloubyansky replied:
Have you tried running the full testsuite with these changes? Or more specifically all the gradle tests including the tooling ones, such add/list extensions, create project? I'm wondering whether we should be changing this method at all. What we need to providing at the end is the AppModel and the workspace structure.

glefloch replied:
I run some tests on my machine on a sample project and everything was ok. Regarding the testsuite, I still have some failures. I will revert that change as you suggest.

aloubyansky commented on devtools/gradle/src/main/java/io/quarkus/gradle/AppModelGradleResolver.java:
```diff
@@ -131,216 +90,65 @@ public Path resolve(AppArtifact appArtifact) throws AppModelResolverException {
+        if (!appArtifact.isResolved()) {
+            PathsCollection.Builder paths = PathsCollection.builder();
+            for (SourceSet sourceSet : model.getSourceSets()) {
+                sourceSet.getSourceDirectories().stream().filter(File::exists).map(File::toPath).forEach(paths::add);
```
Are these java sources?

glefloch replied:
these are more `BuildOutput` such as `build/classes/java/main`

glefloch replied:
well actually it's not, but I will update so it points to classes and not java files

aloubyansky commented on devtools/gradle/src/main/java/io/quarkus/gradle/tasks/QuarkusDev.java:
```diff
@@ -284,7 +284,8 @@ public void startDev() {
             final AppModel appModel;
-            final AppModelResolver modelResolver = extension().getAppModelResolver(LaunchMode.DEVELOPMENT);
+            final QuarkusModel quarkusModel = extension().getQuarkusModel(LaunchMode.DEVELOPMENT);
+            final AppModelResolver modelResolver = new AppModelGradleResolver(quarkusModel);
```
Looks like this could stay the way it was?

aloubyansky commented on devtools/gradle/src/main/java/io/quarkus/gradle/workspace/builder/QuarkusModelBuilder.java:
```diff
@@ -0,0 +1,239 @@
+        }
+
+        final SourceSet mainSourceSet = QuarkusGradleUtils.getSourceSet(project, SourceSet.MAIN_SOURCE_SET_NAME);
+        final SourceSet testSourceSet = QuarkusGradleUtils.getSourceSet(project, SourceSet.TEST_SOURCE_SET_NAME);
```
Looks like this could depend on the launch mode.

aloubyansky commented on independent-projects/bootstrap/app-model/src/main/java/io/quarkus/bootstrap/model/ArtifactKey.java:
```diff
@@ -0,0 +1,13 @@
+package io.quarkus.bootstrap.model;
+
+public interface ArtifactKey {
```
Is this used anywhere else then in AppArtifactKey?

aloubyansky commented on devtools/gradle/src/main/java/io/quarkus/gradle/workspace/model/QuarkusModel.java:
```diff
@@ -0,0 +1,15 @@
+
+    Collection<Dependency> getAppDependencies();
+
+    Collection<String> getExtensions();
```
What is this for? It looks like it also includes all the enforced platforms.

aloubyansky commented on devtools/gradle/src/main/java/io/quarkus/gradle/workspace/builder/QuarkusModelBuilder.java:
```diff
@@ -0,0 +1,239 @@
+        final SourceSet mainSourceSet = QuarkusGradleUtils.getSourceSet(project, SourceSet.MAIN_SOURCE_SET_NAME);
+        final SourceSet testSourceSet = QuarkusGradleUtils.getSourceSet(project, SourceSet.TEST_SOURCE_SET_NAME);
+
+        final Collection<org.gradle.api.artifacts.Dependency> platformExtension = getPlatformExtension(project);
```
I am not sure what platformExtension signifies here. These are application direct extension dependencies, maybe let's call it that?

glefloch replied:
Yes, I agree with you

aloubyansky commented on devtools/gradle/src/main/java/io/quarkus/gradle/workspace/builder/QuarkusModelBuilder.java:
```diff
@@ -0,0 +1,239 @@
+            platformExtension.addAll(getExtensions(configuration.getFirstLevelModuleDependencies(), appDependencies));
+        }
+
+        final Collection<Dependency> platformDependencies = collectExtensionDependencies(project, platformExtension);
```
Let's call them either extension dependencies or deployment dependencies.

glefloch commented:
@aloubyansky thanks for the review, I will update my branch :)

aloubyansky commented on independent-projects/bootstrap/app-model/pom.xml:
```diff
@@ -14,6 +14,22 @@
+        <dependency>
+            <groupId>org.gradle</groupId>
+            <artifactId>gradle-core-api</artifactId>
+            <version>6.1.1</version>
```
I didn't realize we'd introduce Gradle deps in the app-model. This is not a good thing. For Maven we have `quarkus-bootstrap-maven-resolver` that depends on app-model and adds the Maven deps. I think we should do the same for Gradle.

aloubyansky replied:
This should be getting now into the launcher JAR. Have you noticed how much bigger it becomes with the Gradle part?

glefloch replied:
Actually, we don't need this dependency anymore, we only need the `gradle-tooling-api`. But, yes I think externalizing it to a module could be a good idea.

aloubyansky commented on devtools/gradle/src/main/java/io/quarkus/gradle/builder/QuarkusModelBuilder.java:
```diff
@@ -0,0 +1,285 @@
+                project.getBuildDir().getAbsoluteFile(), getSourceSourceSet(mainSourceSet), convert(mainSourceSet));
+    }
+
+    private Set<Workspace> getAdditionalWorkspaces(Project project, String mainProjectName) {
```
There is only one workspace that may include multiple projects.

glefloch replied:
Yes, I will introduce a `project` or `module` type to only keep one `worksapce`

aloubyansky commented on devtools/gradle/src/main/java/io/quarkus/gradle/builder/QuarkusModelBuilder.java:
```diff
@@ -0,0 +1,285 @@
+        return workspaces;
+    }
+
+    private Set<org.gradle.api.artifacts.Dependency> getDirectExtension(Project project) {
```
This method should be renamed. It's not about extensions here. All it does it collects the enforced platforms. It should reflect that in the name.

glefloch replied:
👍 , I will rename it

aloubyansky commented on devtools/gradle/src/main/java/io/quarkus/gradle/builder/QuarkusModelBuilder.java:
```diff
@@ -0,0 +1,285 @@
+            }
+
+            final Dependency dependency = toDependency(a, deploymentConfig.getName());
+            platformDependencies.add(dependency);
```
It looks like you are not checking for potential conflicts with the user dependencies. That's what I used the versionMap for.

glefloch replied:
I though it was solved using the `alignVersion`

aloubyansky replied:
Yes, I found it later in another QuarkusModelHelper.

aloubyansky commented on devtools/gradle/src/main/java/io/quarkus/gradle/builder/QuarkusModelBuilder.java:
```diff
@@ -0,0 +1,285 @@
+    }
+
+    private Set<org.gradle.api.artifacts.Dependency> getDirectExtensionDependencies(Set<ResolvedDependency> dependencies,
+            Collection<Dependency> appDependencies) {
```
1) this method is supposed to collect only the first met extension artifacts (could be direct or transitive but the first met walking the dependency tree;
2) in the original code I was also checking whether I have already visited some deps (that could happen);
3) just fyi, we try to stay away from lambda expressions for now for performance reasons.

aloubyansky commented on devtools/gradle/src/main/java/io/quarkus/gradle/tasks/QuarkusPrepare.java:
```diff
@@ -62,36 +62,47 @@ public void prepareQuarkus() {
             final Convention convention = getProject().getConvention();
             JavaPluginConvention javaConvention = convention.findPlugin(JavaPluginConvention.class);
             if (javaConvention != null) {
+                System.out.println("quarkus prepare java convention found");
```
This should either be removed or logged using getLogger().

glefloch replied:
yes, I will this one

glefloch replied:
and all other from that file ;)

aloubyansky commented on independent-projects/bootstrap/gradle-resolver/src/main/java/io/quarkus/bootstrap/resolver/model/QuarkusModel.java:
```diff
@@ -0,0 +1,15 @@
+
+    Workspace getWorkspace();
+
+    Set<Workspace> getAdditionalWorkspace();
```
As previously mentioned, there should only be one workspace that includes one or more projects or modules.

aloubyansky commented on independent-projects/bootstrap/gradle-resolver/src/main/java/io/quarkus/bootstrap/resolver/model/impl/ArtifactCoordsImpl.java:
```diff
@@ -0,0 +1,52 @@
+import io.quarkus.bootstrap.resolver.model.ArtifactCoords;
+import java.io.Serializable;
+
+public class ArtifactCoordsImpl implements ArtifactCoords, Serializable {
```
equals/hashCode would be good here just to stay safe.

glefloch commented:
@aloubyansky, I updated to pull request. 
I also moved model builder test into the gradle plugin project. when these tests where in `bootstrap/gradle-resolver`, running a `mvn clean install` on quarkus was failing because the gradle plugin was not installed.

aloubyansky commented on devtools/gradle/src/main/java/io/quarkus/gradle/AppModelGradleResolver.java:
```diff
@@ -137,224 +105,11 @@ public Path resolve(AppArtifact appArtifact) throws AppModelResolverException {
     @Override
     public AppModel resolveModel(AppArtifact appArtifact) throws AppModelResolverException {
-        AppModel.Builder appBuilder = new AppModel.Builder();
         if (appModel != null && appModel.getAppArtifact().equals(appArtifact)) {
```
If appModel is not null and the appArtifact does not match, it's an error.

aloubyansky commented on devtools/gradle/src/main/java/io/quarkus/gradle/builder/QuarkusModelBuilder.java:
```diff
@@ -0,0 +1,299 @@
+    public Set<WorkspaceModule> getWorkspace(Project project) {
+        Set<WorkspaceModule> modules = new HashSet<>();
+        if (project.getSubprojects().isEmpty()) {
+            return Collections.singleton(getWorkspaceModule(project));
```
The `project`'s module is not added to the `modules` otherwise.

glefloch replied:
well, I think you right, as long as, it as some java plugin with source, it should be added

aloubyansky commented on devtools/gradle/src/main/java/io/quarkus/gradle/builder/QuarkusModelBuilder.java:
```diff
@@ -0,0 +1,299 @@
+                project.getBuildDir().getAbsoluteFile(), getSourceSourceSet(mainSourceSet), convert(mainSourceSet));
+    }
+
+    private Set<org.gradle.api.artifacts.Dependency> getEnforcedPlatform(Project project) {
```
That's kind of a minor one but it should be plural: platforms :)

aloubyansky commented on devtools/gradle/src/main/java/io/quarkus/gradle/builder/QuarkusModelBuilder.java:
```diff
@@ -0,0 +1,299 @@
+                continue;
+            }
+            final org.gradle.api.artifacts.Dependency deploymentArtifact = getDeploymentArtifact(appDep);
+            if (deploymentArtifact != null) {
```
Again, in this method you are collecting all the extensions, while we should be collecting only the first met ones. Which means if you found an extension, you should not be walking though its dependencies.

glefloch replied:
ok, I now understant the `firstLevel` boolean, I was also missing something on the `visited`. I fixed that

aloubyansky commented on devtools/gradle/src/main/java/io/quarkus/gradle/tasks/QuarkusPrepare.java:
```diff
@@ -74,24 +73,26 @@ public void prepareQuarkus() {
                     throw new GradleException("Failed to find " + INIT_AND_RUN + " method in " + CodeGenerator.class.getName());
                 }
-
+                sourcesDirectories.forEach(System.out::println);
```
Is this a left-over?

glefloch replied:
oupss

aloubyansky commented on independent-projects/bootstrap/gradle-resolver/src/main/java/io/quarkus/bootstrap/resolver/model/Workspace.java:
```diff
@@ -0,0 +1,13 @@
+
+    WorkspaceModule getMainModule();
+
+    Set<WorkspaceModule> getAllModule();
```
This should be plural, i.e. `getAllModules`

glefloch commented:
@aloubyansky, I fixed the gradle windows issue. It was a problem with the `Process` running the wrapper command. 

CI should be ok now :)

aloubyansky commented on bom/application/pom.xml:
```diff
@@ -201,6 +201,7 @@
         <webjars-locator-core.version>0.46</webjars-locator-core.version>
         <log4j2-jboss-logmanager.version>1.0.0.Beta1</log4j2-jboss-logmanager.version>
         <avro.version>1.10.0</avro.version>
+        <gradle-tooling.version>6.5</gradle-tooling.version>
```
Looks like this property could be removed from this pom. But let's see if there are other things to change in this PR first.

aloubyansky commented on devtools/gradle/src/main/java/io/quarkus/gradle/AppModelGradleResolver.java:
```diff
@@ -137,224 +105,15 @@ public Path resolve(AppArtifact appArtifact) throws AppModelResolverException {
-                    }
-                    appArtifact.setPaths(paths.build());
-                }
+                throw new AppModelResolverException("Requested artifact does not match loaded model");
```
It'd be good to include the conflicting artifacts into the message.

aloubyansky commented on devtools/gradle/src/main/java/io/quarkus/gradle/builder/QuarkusModelBuilder.java:
```diff
@@ -0,0 +1,299 @@
+            }
+
+            final Set<ResolvedDependency> resolvedChildren = d.getChildren();
+            if (!resolvedChildren.isEmpty()) {
```
I had `addChildExtension` and walked the whole tree because we also used to collect all the extension properties. Given that you are doing it elsewhere, I think you could have changed this if to `addChildExtension && !resolvedChildren.isEmpty()`. I don't think you need `firstLevel` arg in your impl.

aloubyansky replied:
I.e. you could have removed `addChildExtension` too. We are simply collecting the first met extension dependencies but not its children.

aloubyansky commented on devtools/gradle/src/test/resources/multi-module-project/settings.gradle:
```diff
@@ -0,0 +1,13 @@
+        gradlePluginPortal()
+    }
+    plugins {
+        id 'io.quarkus' version '999-SNAPSHOT'
```
We can't be using specific quarkus versions in the test configs. This will break the release process.

aloubyansky commented on devtools/gradle/src/test/resources/simple-module-project/settings.gradle:
```diff
@@ -0,0 +1,11 @@
+        gradlePluginPortal()
+    }
+    plugins {
+        id 'io.quarkus' version '999-SNAPSHOT'
```
Here as well.

aloubyansky commented on independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/utils/BuildToolHelper.java:
```diff
@@ -0,0 +1,80 @@
+            if (Files.exists(currentPath.resolve(BuildTool.MAVEN.getBuildFile()))) {
+                return true;
+            }
+            currentPath = currentPath.getParent();
```
Did you copy this from somewhere? This is not a good idea. What kind of path is expected to be passed in? Project root or any path inside a project?
A little bit better version would be checking each directory for all the recognizable build files instead of walking up to the root of the FS for each build file.

glefloch replied:
yes I copier that from here: https://github.com/quarkusio/quarkus/blob/42dbdd05cd20ef1e82ad62e6617a48e8ea9100c8/independent-projects/bootstrap/maven-resolver/src/main/java/io/quarkus/bootstrap/resolver/maven/workspace/LocalProject.java#L217

aloubyansky replied:
Ok, that path is derived from a class name (that is in the classes dir of a project) or is a project root.

aloubyansky commented on independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/utils/BuildToolHelper.java:
```diff
@@ -0,0 +1,80 @@
+
+    public static QuarkusModel enableGradleAppModel(Path projectRoot, String mode)
+            throws IOException, AppModelResolverException {
+        if (isMavenProject(projectRoot)) {
```
If projectRoot is actually the project root we shouldn't be walking up to the root of the FS looking for the pom.xml.

aloubyansky commented on independent-projects/bootstrap/gradle-resolver/src/main/java/io/quarkus/bootstrap/resolver/model/impl/WorkspaceImpl.java:
```diff
@@ -0,0 +1,38 @@
+public class WorkspaceImpl implements Workspace, Serializable {
+
+    public ArtifactCoords mainModuleKey;
+    public Set<WorkspaceModule> modules;
```
WorkspaceModuleImpl does not implement hashCode/equals. It would probably be better to have a map instead here. Also given that there is getModule(key)

aloubyansky commented on independent-projects/bootstrap/gradle-resolver/src/main/java/io/quarkus/bootstrap/util/QuarkusModelHelper.java:
```diff
@@ -0,0 +1,179 @@
+        Map<AppArtifactKey, AppDependency> versionMap = new HashMap<>();
+        model.getAppDependencies().stream().map(QuarkusModelHelper::toAppDependency).forEach(appDependency -> {
+            userDeps.add(appDependency);
+            versionMap.put(appArtifact.getKey(), appDependency);
```
you need appDependency.getArtifact().getKey() or something here.

aloubyansky commented on independent-projects/bootstrap/gradle-resolver/src/main/java/io/quarkus/bootstrap/util/QuarkusModelHelper.java:
```diff
@@ -0,0 +1,179 @@
+                    }
+                }
+            }
+            if (!userDeps.contains(appDep)) {
```
This is not correct. You need to be comparing the keys instead. Otherwise, you are not aligning versions here.

glefloch replied:
This is actually done like this in the current model, should i update it to use the `appDep.getArtifact()` instead? (https://github.com/quarkusio/quarkus/blob/23b3d64ead8d67a4ed6e5b88a4853f0a8aaefe20/devtools/gradle/src/main/java/io/quarkus/gradle/AppModelGradleResolver.java#L189)

aloubyansky replied:
Oh, that's clearly a bug. It should be `appDep.getArtifact().getKey()`

aloubyansky commented on integration-tests/gradle/src/test/resources/test-resources-in-build-steps/application/build.gradle:
```diff
@@ -4,6 +4,7 @@ plugins {
 
 dependencies {
     implementation project(":runtime")
+    implementation project(":deployment")
```
No, applications can't depend on the deployment artifact.

aloubyansky commented on test-framework/junit5/src/main/java/io/quarkus/test/junit/QuarkusTestExtension.java:
```diff
@@ -168,6 +170,12 @@ private ExtensionState doJavaStart(ExtensionContext context, Class<? extends Qua
 
+            Path root = Paths.get("").normalize().toAbsolutePath();
+            // If gradle project running directly with IDE
+            if (System.getProperty(BootstrapConstants.SERIALIZED_APP_MODEL) == null) {
```
Where is this property set? I see it in `exportModel` but has it been called by this point?

glefloch replied:
It could already have been set by the gradle plugin in the before test

aloubyansky replied:
We'll need a delegating impl of AppModelResolver. For now it's ok, I think.

aloubyansky reviewed (approved):
Looks good. We'll need to refactor some bootstrap API and setup later to have a cleaner Maven/Gradle bootstrap integration.
